generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -172,6 +173,102 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
+   * -------------------------------------
+   * There are some other things to know about this algorithm:
+   * [0  1  2  3 ]     - When constructing a square matrix in GLSL, the mat constructor takes
+   * [4  5  6  7 ]       matrixDimension * matrixDimension number of arguments - this diagram shows
+   * [8  9  10 11]       how the matrix is constructed from those arguments.
+   * [12 13 14 15]
+   *  -  Notice how in the above diagram, the diagonals for a 4x4 matrix are arguments 0, 5, 10, 15.
+   *     This occurs similarly for 3x3 (0, 4, 8) and 2x2 (0, 3). The sequence of constructor indices
+   *     that are diagonals: [0, (matrixDim + 1), 2 * (matrixDim + 1), 3 * (matrixDim + 1)...].
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from the determinant of a triangular matrix.
+   */
+  private Optional<Expr> opaqueZeroOrOneMatrixDet(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    // TODO(653): Workaround for glslangvalidator. Remove when #653 is fixed.
+    if (constContext) {
+      return Optional.empty();
+    }
+    if (type != BasicType.FLOAT) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedDeterminant()) {
+      return Optional.empty();
+    }
+    // If true, we will make an upper triangular matrix - otherwise it will be lower triangular.
+    final boolean isUpperTriangular = generator.nextBoolean();
+    // Matrices sent to determinant() must be 2x2, 3x3, or 4x4.
+    final int matrixDimension = generator.nextInt(3) + 2;
+    // Indices of arguments in a matrix constructor that can be nonzero for triangular matrix,
+    // excluding the diagonal of the matrix. Zero-indexed.
+    final List<List<Integer>> triangularModifiableArgs =
+        isUpperTriangular
+        ? Arrays.asList(
+            Arrays.asList(1),                     // mat2, upper triangular
+            Arrays.asList(1, 2, 5),               // mat3, upper triangular
+            Arrays.asList(1, 2, 3, 6, 7, 11))     // mat4, upper triangular
+        : Arrays.asList(
+            Arrays.asList(2),                     // mat2, lower triangular
+            Arrays.asList(3, 6, 7),               // mat3, lower triangular
+            Arrays.asList(4, 8, 9, 12, 13, 14));  // mat4, lower triangular
+    // Index of the corresponding triangularModifiableArgs for the given matrix dimension.
+    final int modifiableArgsListIndex = matrixDimension - 2;
```

Comment:
Maybe all of the above would be clearer if you did something like:

List<Integer> modifiableArgs;
switch (matrixDimension) {
  case 2:
    modifiableArgs = isUpperTriangular ? Arrays.asList(...) : Arrays.asList(...);
    break;
  case 3:
    modifiableArgs = isUpperTriangular ? Arrays.asList(...) : Arrays.asList(...);
    break;
  case 4:
    modifiableArgs = isUpperTriangular ? Arrays.asList(...) : Arrays.asList(...);
    break;
  default:
    throw new UnsupportedOperationException(...);
}

That would get rid of the need for the list of lists, and would also avoid the need for "modifiableArgsListIndex" (and its associated very slightly mysterious initialization to matrixDimension - 2).

Reply:
Much more self-documenting, too - thanks.